### PR TITLE
Import tests from std library - part 15/16

### PR DESCRIPTION
**Pull request by SergeyKopienko**
Part 15/16
Please see details in part 1/16 - https://github.com/oneapi-src/oneDPL/pull/885
This PR should be merged into main only after https://github.com/oneapi-src/oneDPL/pull/885

### CODE REVIEW

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                dpl::tuple<int> t3 = dpl::tuple_cat(t1, t2);
+                ret_access[0] &= (dpl::get<0>(t3) == 2);
+            }
+
+            {
+                dpl::tuple<> t1;
+                dpl::tuple<int> t2(2);
+                dpl::tuple<int> t3 = dpl::tuple_cat(t2, t1);
+                ret_access[0] &= (dpl::get<0>(t3) == 2);
+            }
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat.pass.cpp:167]: Seems almost identical to the previous one:

```suggestion
```

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    MoveOnlyData
+    operator+(const MoveOnlyData& x) const
+    {
+        return MoveOnlyData{idx_ + x.idx_};
+    }
+    MoveOnlyData
+    operator*(const MoveOnlyData& x) const
+    {
+        return MoveOnlyData{idx_ * x.idx_};
+    }
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat.pass.cpp:57]: I have looked through the test and it seems like these functions are not necessary

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat1_g.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
+// -*- C++ -*-
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat1_g.pass.cpp:1]: I guess all of the type checks are done in `tuple_cat.pass.cpp` and this test can be removed.

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_element_g.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+{
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        cgh.single_task<class KernelTest1>([=]() {
+            {
+                static_assert(
+                    dpl::is_same<dpl::tuple_element<0, const dpl::tuple<float, void, int>>::type, const float>::value);
+                static_assert(
+                    dpl::is_same<dpl::tuple_element<1, volatile dpl::tuple<short, void>>::type, volatile void>::value);
+                static_assert(dpl::is_same<dpl::tuple_element<2, const volatile dpl::tuple<float, char, int>>::type,
+                                           const volatile int>::value);
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_element_g.pass.cpp:37]: Can we improve readability and improve coverage here:

```suggestion
                using tuple_type = dpl::tuple<float, short, int>;
                static_assert(dpl::is_same_v<dpl::tuple_element<0, const tuple_type>::type, const float>);
                static_assert(dpl::is_same_v<dpl::tuple_element<1, volatile tuple_type>::type, volatile short>);
                static_assert(dpl::is_same_v<dpl::tuple_element<2, const volatile tuple_type>::type, const volatile int>);
                
                static_assert(dpl::is_same_v<dpl::tuple_element_t<0, const tuple_type>, const float>;
                static_assert(dpl::is_same_v<dpl::tuple_element_t<1, volatile tuple_type>, volatile short>);
                static_assert(dpl::is_same_v<dpl::tuple_element_t<2, const volatile tuple_type>, const volatile int>);
```

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_element_g.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    });
+}
+void
+kernel_test2(sycl::queue& deviceQueue)
+{
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        cgh.single_task<class KernelTest2>([=]() {
+            {
+                static_assert(dpl::is_same<dpl::tuple_element<0, const dpl::tuple<double, void, int>>::type,
+                                           const double>::value);
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_element_g.pass.cpp:49]:
```suggestion
                using tuple_type = dpl::tuple<double, void, int>;
                static_assert(dpl::is_same_v<dpl::tuple_element<0, const tuple_type>::type, const double>);
                static_assert(dpl::is_same_v<dpl::tuple_element_t<0, const tuple_type>, const double>);
```

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_size_g.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
+// -*- C++ -*-
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_size_g.pass.cpp:1]: I don't thing it is important to test tuple_size with different cv-qualifiers of tuple as the impact is insignificant. I propose to remove this test.

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_rv.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+{
+    sycl::queue deviceQueue = TestUtils::get_test_queue();
+    bool ret = false;
+    sycl::range<1> numOfItems{1};
+    sycl::buffer<bool, 1> buffer1(&ret, numOfItems);
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        auto ret_access = buffer1.get_access<sycl::access::mode::write>(cgh);
+        cgh.single_task<class KernelGetConstRvTest>([=]() {
+            {
+                dpl::tuple<int> t(3);
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_rv.pass.cpp:39]: Since this is the test for `get(const tuple<Types...>&&)` this variable should be constant.
Some checks are required to be changed as well.

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_tuple_elem.pass.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    sycl::queue deviceQueue = TestUtils::get_test_queue();
+    bool ret = false;
+    sycl::range<1> numOfItems{1};
+    sycl::buffer<bool, 1> buffer1(&ret, numOfItems);
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        auto ret_access = buffer1.get_access<sycl::access::mode::write>(cgh);
+        cgh.single_task<class KernelGetConstTest>([=]() {
+            {
+                typedef dpl::tuple<int> T;
+                const T t(3);
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_tuple_elem.pass.cpp:42]:
```suggestion
                const dpl::tuple<int> t(3);
```

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_tuple_elem.pass.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "support/test_macros.h"
+#include "support/utils.h"
+
+#if TEST_DPCPP_BACKEND_PRESENT
+class KernelGetConstTest;
+
+struct Empty
+{
+};
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_tuple_elem.pass.cpp:28]: Seems unused.

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_non_const.pass.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    sycl::queue deviceQueue = TestUtils::get_test_queue();
+    bool ret = false;
+    sycl::range<1> numOfItems{1};
+    sycl::buffer<bool, 1> buffer1(&ret, numOfItems);
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        auto ret_access = buffer1.get_access<sycl::access::mode::write>(cgh);
+        cgh.single_task<class KernelGetNonConstTest>([=]() {
+            {
+                typedef dpl::tuple<int> T;
+                T t(3);
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/get_non_const.pass.cpp:57]:
```suggestion
                dpl::tuple<int> t(3);
```

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_non_const.pass.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            }
+
+            { // get on an rvalue tuple
+                static_assert(dpl::get<0>(dpl::make_tuple(0.0f, 1, 2.0, 3L)) == 0);
+                static_assert(dpl::get<1>(dpl::make_tuple(0.0f, 1, 2.0, 3L)) == 1);
+                static_assert(dpl::get<2>(dpl::make_tuple(0.0f, 1, 2.0, 3L)) == 2);
+                static_assert(dpl::get<3>(dpl::make_tuple(0.0f, 1, 2.0, 3L)) == 3);
+                static_assert(S().k == 1);
+                static_assert(dpl::get<1>(getP()) == 4);
+            }
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/get_non_const.pass.cpp:70]: It seems like these tests are not related to `get(tuple<Types>&)`. So I guess it should be removed together with helpers.

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_rv.pass.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    sycl::range<1> numOfItems{1};
+    sycl::buffer<bool, 1> buffer1(&ret, numOfItems);
+    deviceQueue.submit([&](sycl::handler& cgh) {
+        auto ret_access = buffer1.get_access<sycl::access::mode::write>(cgh);
+        cgh.single_task<class KernelGetRvTest>([=]() {
+            int test_val = 3;
+            typedef dpl::tuple<int*> T;
+            T t(&test_val);
+            int* p = dpl::get<0>(std::move(t));
+            ret_access[0] = (*p == 3);
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.elem/get_rv.pass.cpp:41]:
```suggestion
            int test_val = 3;
            dpl::tuple<int*> t(&test_val);
            int* p = dpl::get<0>(std::move(t));
            ret_access[0] = (*p == 3);
```

**Comment by kboyarinov** [comment]
Also, I propose to rename some tests:

tuple.creation
     `constexpr_g.pass.cpp` -> `non_member_constexpr.pass.cpp`
     `tie.pass.cpp` -> `tuple_tie1.pass.cpp`
     `tie1_g.pass.cpp` -> `tuple_tie2.pass.cpp`
     `tie2_g.pass.cpp` -> `tuple_tie_pair.pass.cpp`
     `tuple_cat.pass.cpp` -> `tuple_cat1.pass.cpp`
     `tuple_cat2_g.pass.cpp` -> `tuple_cat2.pass.cpp`
tuple.elem
     `cv_tuple_element_g.pass.cpp` -> `cv_tuple_element.pass.cpp`
     `get_const_rv.pass.cpp` -> `tuple_get_const_rvalue.pass.cpp`
     `get_const_tuple_elem.pass.cpp` -> `tuple_get_const_lvalue.pass.cpp`
     `get_non_const.pass.cpp` -> `tuple_get_lvalue.pass.cpp`
     `get_rv.pass.cpp` -> `tuple_get_rvalue.pass.cpp`

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/forward_as_tuple1_g.pass.cpp [hunk 13]
```diff
@@ -0,0 +1,111 @@
+// -*- C++ -*-
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/forward_as_tuple1_g.pass.cpp:1]: Done

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/forward_as_tuple2_g.pass.cpp [hunk 14]
```diff
@@ -0,0 +1,52 @@
+// -*- C++ -*-
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/forward_as_tuple2_g.pass.cpp]: Done

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tie.pass.cpp [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include <oneapi/dpl/tuple>
+#include <oneapi/dpl/functional>
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/tie.pass.cpp:19]: Removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/make_tuple.pass.cpp [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include <oneapi/dpl/tuple>
+#include <oneapi/dpl/functional>
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/make_tuple.pass.cpp]: Removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tie1_g.pass.cpp [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include <oneapi/dpl/tuple>
+#include <oneapi/dpl/utility>
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/tie1_g.pass.cpp]: Fixed

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_tuple_elem.pass.cpp (hunk 9 above)
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_tuple_elem.pass.cpp]: Removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat.pass.cpp (hunk 2 above)
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat.pass.cpp]: Fixed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat1_g.pass.cpp (hunk 3 above)
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/tuple_cat1_g.pass.cpp]: Done: removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/make_tuple1_g.pass.cpp [hunk 18]
```diff
@@ -0,0 +1,54 @@
+// -*- C++ -*-
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/make_tuple1_g.pass.cpp:1]: Done: removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/make_tuple2_g.pass.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    bool ret = false;
+    sycl::range<1> numOfItems{1};
+    {
+        sycl::buffer<bool, 1> buffer1(&ret, numOfItems);
+        deviceQueue.submit([&](sycl::handler& cgh) {
+            auto ret_access = buffer1.get_access<sycl::access::mode::write>(cgh);
+            cgh.single_task<class KernelTest>([=]() {
+                {
+                    auto x = A{7};
+                    dpl::make_tuple(x);
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/make_tuple2_g.pass.cpp]: Done: removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/constexpr_g.pass.cpp [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+        constexpr tuple_type t1{55, 77.77f};
+        [[maybe_unuased]] constexpr auto var = dpl::get<1>(t1);
+    }
+
+    {
+        typedef dpl::tuple<int, float, int> tuple_type;
+        constexpr tuple_type t1{55, 77.77f, 99};
+        [[maybe_unuased]] constexpr auto var = dpl::get<2>(t1);
+    }
+}
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/constexpr_g.pass.cpp]: Done: function removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/constexpr_g.pass.cpp [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace
+{
+template <class T>
+constexpr int zero_from_anything(T)
+{
+    return 0;
+}
+} // namespace
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/constexpr_g.pass.cpp]: Done: removed

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_size_g.pass.cpp (hunk 6 above)
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.elem/cv_tuple_size_g.pass.cpp]: Done: test removed

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/constexpr_g.pass.cpp [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    [[maybe_unuased]] constexpr int ign3 = zero_from_anything(dpl::ignore);
+}
+
+void
+kernel_test()
+{
+    sycl::queue deviceQueue = TestUtils::get_test_queue();
+    {
+        deviceQueue.submit([&](sycl::handler& cgh) {
+            cgh.single_task<class KernelTest>([=]() {
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.creation/constexpr_g.pass.cpp:118]: Done

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_non_const.pass.cpp (hunk 11 above)
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.elem/get_non_const.pass.cpp]: Done

File: test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_rv.pass.cpp (hunk 7 above)
Comment (SergeyKopienko) [reply test/xpu_api/tuple/tuple.tuple/tuple.elem/get_const_rv.pass.cpp]: Done

File: test/xpu_api/tuple/tuple.tuple/tuple.creation/forward_as_tuple.pass.cpp [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+    static_assert(dpl::is_same<typename dpl::tuple_element<1, Tuple>::type, char&>::value);
+    return (dpl::get<0>(t) == 2.5f && dpl::get<1>(t) == 'a');
+}
+
+template <class Tuple>
+constexpr int
+test3(const Tuple&)
+{
+    return dpl::tuple_size<Tuple>::value;
+}
```
Comment (kboyarinov) [inline test/xpu_api/tuple/tuple.tuple/tuple.creation/forward_as_tuple.pass.cpp:59]: Seems not used anymore:
```suggestion
```

**Review by kboyarinov — APPROVED** [review]
